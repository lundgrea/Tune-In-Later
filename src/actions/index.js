export const addAlbums = albums => ({
  type: "ADD_ALBUMS",
  albums
});

export const login = user => ({
  type: "LOGIN",
  user
});

export const logout = () => ({
  type: "LOGOUT"
});

export const toggleFavorite = (id) => ({
  type: "TOGGLE_FAVORITE",
  id
});

export const createUser = user => ({
  type: "CREATE_USER",
  user
});

export const handleErrors = error => ({
  type: "HANDLE_ERROR",
  error
});

export const storeFavorites = albums => ({
  type: "STORE_FAVORITES",
  albums
});

// export const deleteFavorite = albumId => ({
//   type: "DELETE_FAVORITE",
//   albumId
// });
