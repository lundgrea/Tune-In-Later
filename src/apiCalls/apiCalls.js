export const getAlbums = async artist => {
  try {
    const response = await fetch(
      `https://itunes.apple.com/search?term=${artist}&entity=album`
    );
    const data = await response.json();
    return data;
  } catch (error) {
    console.log(error.message);
    throw new Error("Failed to get albums");
  }
};
export const loginUser = async user => {
  try {
    const options = {
      method: "POST",
      body: JSON.stringify(user),
      headers: {
        "Content-Type": "application/json"
      }
    };
    const response = await fetch("http://localhost:3001/api/v1/login", options);
    const result = await response.json();
    return result;
  } catch (error) {
    throw new Error("Incorrect login information");
  }
};

export const postUser = async newUser => {
  console.log("newUser :", newUser);
  try {
    const options = {
      method: "POST",
      body: JSON.stringify(newUser),
      headers: {
        "Content-Type": "application/json"
      }
    };
    const response = await fetch("http://localhost:3001/api/v1/users", options);
    const result = await response.json();
    console.log("result :", result);
    return result;
  } catch (error) {
    throw new Error(error);
  }
};

export const postFavorite = async (user, newFavorite) => {
  console.log('newFavorite :', newFavorite);
  try {
    const options = {
      method: "POST", 
      body: JSON.stringify(newFavorite),
      headers: {
        'Content-Type': 'application/json'
      }
    }
    const response = await fetch(`http://localhost:3001/api/v1/users/${user}/albumfavorites`, options)
    const result = await response.json();
    return result;
  } catch (error) {
    throw new Error (error);
  }
}