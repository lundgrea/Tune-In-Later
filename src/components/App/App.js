import React from "react";
import Search from "../../containers/Search/Search";
import "./App.css";
import { Route, NavLink } from "react-router-dom";
import CardContainer from "../../containers/CardContainer/CardContainer";
import LoginForm from "../LoginForm/LoginForm";
import { SignUpForm } from "../SignUpForm/SignUpForm";

const App = () => {
  return (
    <>
      <Route exact path="/" />
      <h1>TUNE-IN LATER</h1>
      <NavLink exact to="/login">
        LOGIN
      </NavLink>
      <Search />
      <CardContainer />
      <SignUpForm />
      <Route exact path="/login" render={() => <LoginForm />} />
    </>
  );
};

export default App;
