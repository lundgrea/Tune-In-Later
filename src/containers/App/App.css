@font-face {
  /* font-family: "Lato"; */
  /* src: url("https://fonts.googleapis.com/css?family=Lato:300,400,900&display=swap">); */
  font-family: "Source Sans Pro";
  src: url("https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700&display=swap")
}

.App {
  text-align: center;
}
body {
  height: 100vh;
  width: 100%;
  margin: 0;
  /* display: grid;
  grid-template-areas: 
    "header header header"
    "aside main main";
  grid-template-columns: 1fr; */
  /* grid-template-rows: 20% auto; */
}

/* ul { 
  list-style-type: none; 
  /* display: inline; */

.login-div{
  margin: 30px 10px 10px 300px;
}

li:hover {
  background-color: FFC602;
}


.nav-links {
  font-size: 20px;
  display: inline;
  color: #a8d0e6;
  float: right;
  margin: 70px 20px 70px 20px;
}

a {
  text-decoration: none;
}


.a-search{
  font-size: 50px;
  color: #a8d0e6;
}

a:focus {
  text-decoration: none;
  color: lightgrey;
}


header{
  background-color: white
}
/* .App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
} */

.logo {
  width: 100px;
  color: #F2B54a
}

.logo-lockup {
  display: flex;
  margin: 10px 10px 10px 10px;
}

h1 {
  color: #FFC602;
  font-size: 60px;
  margin: 62px 10px 10px 30px;
}

.App-link {
  color: #61dafb;
}

/* @keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
} */

nav {
  max-height: 30vh;
  width: 100%;
  display: flex;
  /* align-items: center; */
  /* justify-content: flex-end; */
  padding-right: 30px;
  background-color: white;
    /* margin-left: 1%; */
    /* display: flex; */
    /* justify-content: flex-end; */
    /* padding-right: 30px; */    
}


/* background-color: #23305E; */








/* .home-nav, .login-ul { */
  /* min-height: 25vh; */
  /* display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly; */
/* } */

.search-form {
  min-height: 100vh;
  width: 27%;
  background-color: #39424E;
  /* border-radius: 10px; */
}

.login-form {
 max-height: 30vh;
 width: 20%;
 background-color: yellow;
 border-radius: 10px;
 display: flex;
 flex-direction: column;
 justify-content: space-around;
 align-items: center;
}

main {
  display: flex;
}

section {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.favorites-container {
  overflow: hidden;
  width: 100%;
  min-height: 60vh;
  display: flex;
  flex-wrap: wrap;
}

.card-container {
  min-height: 30vh;
  width: 30vh;
  flex: 1 0 25%;
  border: grey solid 2px;
  border-radius: 8px;
  margin: 15px;
  text-align: center;
}