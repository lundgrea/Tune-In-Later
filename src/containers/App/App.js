import React from "react";
import Search from "../Search/Search";
import "./App.css";
import { getFavorites } from "../../apiCalls/apiCalls";
import { Route, NavLink, Link } from "react-router-dom";
import CardContainer from "../CardContainer/CardContainer";
import LoginForm from "../LoginForm/LoginForm";
import SignUpForm from "../SignUpForm/SignUpForm";
import { logout, storeFavorites } from '../../actions';
import { connect } from 'react-redux';
import  FavoriteContainer from "../Favorites/FavoriteContainer";
import { AlbumDetails } from "../../Components/AlbumDetails/AlbumDetails";
import PropTypes from 'prop-types'




export const App = ({ logout, error, user, storeFavorites, albums}) => {
  const sendFavorites = async () => {
    const something = await getFavorites(user.id)
    const newFavorites = something.map(album => ({
      ...album, isFavorite: true
    }))
    storeFavorites(newFavorites)
  } 
  return (
  <main>
    <form className='search-form'>
      <h1>TUNE-IN LATER</h1>
        <NavLink exact to='/search'>Search</NavLink>
        {user && <NavLink exact to='/' onClick={logout}>LOGOUT</NavLink>}
      <Route 
      exact path='/search' 
      component={Search}/>
      </form>
      <section>
      <nav>
        <p>{error}</p>
        <div className='login-div'>
          {!user && <Link  to='/sign-up'><button>Sign Up Page</button></Link>}
          {!user && <Link  to="/login"><button>Login Page</button></Link>}
        </div>
        <div className='home-nav'>
          {user && <NavLink exact to='/my-collection' onClick={sendFavorites}>Favorites</NavLink>}
          {user && <NavLink exact to='/'>Home</NavLink>}
        </div>
      </nav>
      <Route exact path="/" />
      <Route 
      exact path="/login" 
      component={LoginForm} 
      />
      <Route 
      exact path='/sign-up' 
      component={SignUpForm} 
      />
      <div className='favorites-container'>
      <Route 
      exact path='/my-collection' 
      component = {FavoriteContainer} 
      />
      <Route 
      exact path='/search'
      component= {CardContainer}
      />
      <Route exact path='/:id' render={({match}) => {
        const {id} = match.params;
        const description = albums.find(album => {
          return album.album_id === parseInt(id)
        })
        return description && <AlbumDetails album={description} />
      }}/>
      </div>
      </section>
    </main>
  );
};




export const mapStateToProps = store => ({
  albums: store.albums,
  error: store.error,
  user: store.user,
});


export const mapDispatchToProps = (dispatch) => ({
  logout: () => dispatch(logout()),
  storeFavorites: (favorites) => dispatch(storeFavorites(favorites))
});

export default connect(mapStateToProps, mapDispatchToProps)(App);


App.propTypes = {
  albums: PropTypes.array.isRequired,
  error: PropTypes.string.isRequired,
  user: PropTypes.object.isRequired,
  logout: PropTypes.func.isRequired,
  storeFavorites: PropTypes.func.isRequired
}