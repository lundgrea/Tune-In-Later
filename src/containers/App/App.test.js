import React from 'react';
import { shallow, mount } from 'enzyme'
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { MemoryRouter } from 'react-router';
import { logout, storeFavorites } from '../../actions';
import { getFavorites } from "../../apiCalls/apiCalls";


describe('App', () => {
  let wrapper;
  const mockAlbums = [{trackName: 'Hello', genre: 'Rock'}, {trackName: 'Yellow', genre: 'Folk'}];
  const mockLogOut = jest.fn();
  const mockStoreFavorites = jest.fn();

  beforeEach(() => {
    wrapper = shallow (<App
      logout={mockLogOut}
      error={"error"}
      user={{name: 'lola', id: 7}}
      storeFavorites={mockStoreFavorites}
      albums={mockAlbums}
    />
    )
  })

  it('should map the snapshot with all the props passed in correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })
})

describe('mapStateToProps', ()  => {
  it('should return an array of albums', () => {
    const mockState = {
      albums: [{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}],
      error: 'error',
      user: 1
    }
    const expected = {
      albums: [{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}],
      error: 'error',
      user: 1
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected)
  })

  it('should return an error if applicable', () => {
    const mockState = {
      albums: [{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}],
      error: 'error',
      user: 1
    }
    const expected = {
      albums: [{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}],
      error: 'error',
      user: 1
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected)
  })
  
  it('should have a user saved if a user is logged in', () => {
    const mockState = {
      albums: [{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}],
      error: 'error',
      user: 1
    }
    const expected = {
      albums: [{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}],
      error: 'error',
      user: 1
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected)
  })
})

describe('mapDispatchToProps', () => {
  it('calls logout with an logout action when logout is called', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = logout();
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.logout();
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('calls storeFavorites with an array of favorites when storeFavorites is called', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = storeFavorites([{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}]);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.storeFavorites([{albumName: 'Johanna', genre: 'Rock'}, {albumName: 'Eoah', genre: 'Country'}]);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
})


describe("Router", () => {
  it("should show the Main page when nothing is selected", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/' ]}>
        <App />
      </MemoryRouter>
    )
    expect(wrapper.find(App)).toHaveLength(1);        
  })
})

  // it("should show the card page when Search is selected", () => {
  //   const mockAlbums = [{trackName: 'Hello', genre: 'Rock'}, {trackName: 'Yellow', genre: 'Folk'}];
  //   const mockLogOut = jest.fn();
  //   const mockStoreFavorites = jest.fn();
  //   const wrapper = mount(
  //     <Provider>
  //     <MemoryRouter initialEntries={[ '/sign-up' ]}>
  //       <App
  //         logout={mockLogOut}
  //         error={"error"}
  //         user={{name: 'lola', id: 7}}
  //         storeFavorites={mockStoreFavorites}
  //         albums={mockAlbums}
  //       />
  //     </MemoryRouter>
  //     </Provider>
  //   )
  //   expect(wrapper.find(SignUpForm)).toHaveLength(1);        
  // })
