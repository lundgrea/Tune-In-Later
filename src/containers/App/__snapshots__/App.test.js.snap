// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should map the snapshot with all the props passed in correctly 1`] = `
<main>
  <form
    className="search-form"
  >
    <h1>
      TUNE-IN LATER
    </h1>
    <NavLink
      exact={true}
      to="/search"
    >
      Search
    </NavLink>
    <NavLink
      exact={true}
      onClick={[MockFunction]}
      to="/"
    >
      LOGOUT
    </NavLink>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Search)",
          "type": [Function],
        }
      }
      exact={true}
      path="/search"
    />
  </form>
  <section>
    <nav>
      <p>
        error
      </p>
      <div
        className="login-div"
      />
      <div
        className="home-nav"
      >
        <NavLink
          exact={true}
          onClick={[Function]}
          to="/my-collection"
        >
          Favorites
        </NavLink>
        <NavLink
          exact={true}
          to="/"
        >
          Home
        </NavLink>
      </div>
    </nav>
    <Route
      exact={true}
      path="/"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(LoginForm)",
          "type": [Function],
        }
      }
      exact={true}
      path="/login"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(SignUpForm)",
          "type": [Function],
        }
      }
      exact={true}
      path="/sign-up"
    />
    <div
      className="favorites-container"
    >
      <Route
        component={
          Object {
            "$$typeof": Symbol(react.memo),
            "WrappedComponent": [Function],
            "compare": null,
            "displayName": "Connect(FavoriteContainer)",
            "type": [Function],
          }
        }
        exact={true}
        path="/my-collection"
      />
      <Route
        component={
          Object {
            "$$typeof": Symbol(react.memo),
            "WrappedComponent": [Function],
            "compare": null,
            "displayName": "Connect(CardContainer)",
            "type": [Function],
          }
        }
        exact={true}
        path="/search"
      />
      <Route
        exact={true}
        path="/:id"
        render={[Function]}
      />
    </div>
  </section>
</main>
`;
